File: PROGRAMMING FUNDAMENTALS/Data Structures/stack.cpp
#include <iostream>
#define MAX 100
using namespace std; 

int stack[MAX];
int top = -1;

// Push operation
void push(int value) {
    if (top >= MAX - 1) {
        std::cout << "Stack Overflow\n";
    } else {
        top = top + 1; 
        stack[top] = value;
        std::cout << value << " pushed to stack\n";
    }
}

// Pop operation
int pop() {
    if (top < 0) {
        std::cout << "Stack Underflow\n";
        return -1;
    } else {
        return stack[top--];
    }
}

// Peek operation
int peek() {
    if (top < 0) {
        std::cout << "Stack is empty\n";
        return -1;
    } else {
        return stack[top];
    }
}

// Check if empty
bool isEmpty() {
    return top == -1;
}

int main() {
    push(10);
    push(20);
    push(30);

    std::cout << "Top element: " << peek() << "\n";
    std::cout << "Popped element: " << pop() << "\n";
    std::cout << "Is stack empty? " << (isEmpty() ? "Yes" : "No") << "\n";

    return 0;
}

File: Python/DATA STRUCTURE/practice/stack.py
# Stack ka maximum size define kar rahe hain
MAX = 100

# Stack ko ek list ke form mein initialize kar rahe hain
stack = []

# Push operation: stack mein value add karna
def push(value):
    if len(stack) >= MAX:
        print("Stack Overflow")  # Agar stack full ho gaya hai toh overflow message
    else:
        stack.append(value)  # Value ko stack ke end mein daal rahe hain
        print(f"{value} pushed to stack")  # Confirmation message

# Pop operation: stack se top value hataana
def pop():
    if len(stack) == 0:
        print("Stack Underflow")  # Agar stack empty hai toh underflow message
        return -1
    else:
        return stack.pop()  # Stack se top element hata ke return kar rahe hain

# Peek operation: stack ka top element dekhna bina hataaye
def peek():
    if len(stack) == 0:
        print("Stack is empty")  # Agar stack empty hai toh message
        return -1
    else:
        return stack[-1]  # Stack ka last element return kar rahe hain

# Check karna ki stack empty hai ya nahi
def isEmpty():
    return len(stack) == 0  # Agar length zero hai toh stack empty hai

# Main function jahan operations perform kar rahe hain
push(10)  # 10 ko stack mein daal rahe hain
push(20)  # 20 ko stack mein daal rahe hain
push(30)  # 30 ko stack mein daal rahe hain

print("Top element:", peek())  # Stack ka top element print kar rahe hain
print("Popped element:", pop())  # Top element ko hata ke print kar rahe hain
print("Is stack empty?", "Yes" if isEmpty() else "No")  # Stack empty hai ya nahi check kar rahe hain


